File: app/src/main/java/com/vmenon/mpo/activity/BaseDrawerCollapsingToolbarActivity.java
package com.vmenon.mpo.activity;

import android.os.Bundle;
import androidx.annotation.Nullable;
import android.view.View;
import android.view.ViewStub;

import com.google.android.material.appbar.AppBarLayout;
import com.google.android.material.appbar.CollapsingToolbarLayout;
import com.google.android.material.floatingactionbutton.FloatingActionButton;
import com.vmenon.mpo.R;

public abstract class BaseDrawerCollapsingToolbarActivity extends BaseDrawerActivity
        implements AppBarLayout.OnOffsetChangedListener {

    private FloatingActionButton fab;
    private CollapsingToolbarLayout collapsingToolbar;
    private boolean collapsed = false;
    private int scrollRange = -1;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        fab = findViewById(R.id.fab);
        fab.setImageResource(getFabDrawableResource());
        fab.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                onFabClick();
            }
        });

        collapsingToolbar = findViewById(R.id.collapsing_toolbar);
        final AppBarLayout appBarLayout = findViewById(R.id.appbar);
        appBarLayout.addOnOffsetChangedListener(this);
    }

    protected abstract int getFabDrawableResource();
    protected abstract CharSequence getCollapsedToolbarTitle();
    protected abstract CharSequence getExpandedToolbarTitle();
    protected abstract void onFabClick();
    protected abstract int getCollapsiblePanelContentLayoutId();

    @Override
    protected int getRootLayoutResourceId() {
        return R.layout.activity_base_drawer_collapsing_toolbar;
    }

    @Override
    protected void inflateContent() {
        ViewStub viewStub = findViewById(R.id.contentViewStub);
        viewStub.setLayoutResource(getLayoutResourceId());
        viewStub.inflate();

        ViewStub panelViewStub = findViewById(R.id.collapsiblePanelViewStub);
        panelViewStub.setLayoutResource(getCollapsiblePanelContentLayoutId());
        panelViewStub.inflate();
    }

    @Override
    public void onOffsetChanged(AppBarLayout appBarLayout, int verticalOffset) {
        if (scrollRange == -1) {
            scrollRange = appBarLayout.getTotalScrollRange();
        }
        if (scrollRange + verticalOffset == 0) {
            collapsingToolbar.setTitle(getCollapsedToolbarTitle());
            collapsed = true;
        } else if (collapsed) {
            collapsingToolbar.setTitle(getExpandedToolbarTitle());
            collapsed = false;
        }
    }
}

File: app/src/main/java/com/vmenon/mpo/activity/BaseDrawerCollapsingToolbarActivity.kt
package com.vmenon.mpo.activity

import android.os.Bundle
import android.view.ViewStub

import com.google.android.material.appbar.AppBarLayout
import com.google.android.material.appbar.CollapsingToolbarLayout
import com.google.android.material.floatingactionbutton.FloatingActionButton
import com.vmenon.mpo.R

abstract class BaseDrawerCollapsingToolbarActivity : BaseDrawerActivity(), AppBarLayout.OnOffsetChangedListener {

    private lateinit var fab: FloatingActionButton
    private lateinit var collapsingToolbar: CollapsingToolbarLayout
    private var collapsed = false
    private var scrollRange = -1

    protected abstract val fabDrawableResource: Int
    protected abstract val collapsedToolbarTitle: CharSequence
    protected abstract val expandedToolbarTitle: CharSequence
    protected abstract val collapsiblePanelContentLayoutId: Int

    override val rootLayoutResourceId: Int
        get() = R.layout.activity_base_drawer_collapsing_toolbar

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        fab = findViewById(R.id.fab)
        fab.setImageResource(fabDrawableResource)
        fab.setOnClickListener { onFabClick() }

        collapsingToolbar = findViewById(R.id.collapsing_toolbar)
        val appBarLayout = findViewById<AppBarLayout>(R.id.appbar)
        appBarLayout.addOnOffsetChangedListener(this)
    }

    protected abstract fun onFabClick()

    override fun inflateContent() {
        val viewStub = findViewById<ViewStub>(R.id.contentViewStub)
        viewStub.layoutResource = layoutResourceId
        viewStub.inflate()

        val panelViewStub = findViewById<ViewStub>(R.id.collapsiblePanelViewStub)
        panelViewStub.layoutResource = collapsiblePanelContentLayoutId
        panelViewStub.inflate()
    }

    override fun onOffsetChanged(appBarLayout: AppBarLayout, verticalOffset: Int) {
        if (scrollRange == -1) {
            scrollRange = appBarLayout.totalScrollRange
        }
        if (scrollRange + verticalOffset == 0) {
            collapsingToolbar.title = collapsedToolbarTitle
            collapsed = true
        } else if (collapsed) {
            collapsingToolbar.title = expandedToolbarTitle
            collapsed = false
        }
    }
}
